src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,176 @@
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(this.OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext startContext)
+        {
+            var stringSymbol = WellKnownTypes.String(startContext.Compilation);
+
+            if (stringSymbol is null)
+            {
+                return;
+            }
+
+            var substringMethods = stringSymbol.GetMembers(nameof(string.Substring)).OfType<IMethodSymbol>();
+            var substring1ParameterMethod = substringMethods
+                .SingleOrDefault(substring => substring.Parameters.Length == 1
+                                    && substring.Parameters[0].Type.SpecialType == SpecialType.System_Int32);
+
+            if (substring1ParameterMethod is null)
+            {
+                return;
+            }
+
+            var substring2ParameterMethod = substringMethods
+                .SingleOrDefault(substring => substring.Parameters.Length == 2
+                                              && substring.Parameters[0].Type.SpecialType == SpecialType.System_Int32
+                                              && substring.Parameters[1].Type.SpecialType == SpecialType.System_Int32);
+
+            if (substring2ParameterMethod is null)
+            {
+                return;
+            }
+
+            var stringBuilderSymbol = WellKnownTypes.StringBuilder(startContext.Compilation);
+
+            if (stringBuilderSymbol is null)
+            {
+                return;
+            }
+
+            var sourceAppendMethod = stringBuilderSymbol
+                .GetMembers(nameof(StringBuilder.Append))
+                .OfType<IMethodSymbol>()
+                .Where(append => append.Parameters.Length == 1)
+                .FirstOrDefault(append => append.Parameters[0].Type.SpecialType == SpecialType.System_String);
+
+            if (sourceAppendMethod is null)
+            {
+                return;
+            }
+            
+            startContext.RegisterOperationAction(
+                context =>
+                {
+                    if (context.Operation is IInvocationOperation invocation)
```

Comment:
>context.Operation is IInvocationOperation invocation) [](start = 24, length = 53)

Replace this with a cast `var invocation = (IInvocationOperation)context.Operation;`. This is guaranteed by the API.